server/src/main/java/com/saralein/server/request/RequestParser.java

```diff
@@ -1,69 +1,71 @@
         try {
-            List<String> fullRequest = Arrays.asList(split(request, CRLF));
-            HashMap<String, String> parsedRequest = createRequestMap(fullRequest);
-            return new Request(parsedRequest);
-        } catch(ArrayIndexOutOfBoundsException e) {
+            String requestLine = request.get(0);
+            return requestLine.split(" ");
+        } catch (ArrayIndexOutOfBoundsException e) {
             throw new Exception("Bad request. Connection closed.");
         }
     }
 
-    private HashMap<String, String> createRequestMap(List<String> fullRequest) {
-        HashMap<String, String> parsedRequest = addRequestLine(fullRequest);
-
-        if (fullRequest.size() > 0) {
-            parsedRequest.putAll(addHeadersAndBody(fullRequest));
-        }
-
-        return parsedRequest;
+    private Request buildRequest(String[] requestLine, List<String> request) {
+        return new Request.Builder()
+                .method(getMethod(requestLine))
+                .uri(getUri(requestLine))
+                .body(parseBody(request))
+                .addHeaders(parseHeaders(request))
+                .build();
     }
 
-    private HashMap<String, String> addRequestLine(List<String> fullRequest) {
-        String[] requestLine = splitRequestLine(fullRequest);
-
-        return new HashMap<String, String>(){{
-            put("method", parseMethod(requestLine));
-            put("uri", parseUri(requestLine));
-            put("version", parseVersion(requestLine));
-        }};
+    private String getMethod(String[] requestLine) {
+        return requestLine[0];
     }
 
-    private HashMap<String, String> addHeadersAndBody(List<String> fullRequest) {
-        HashMap<String, String> headersAndBody = new HashMap<>();
+    private String getUri(String[] requestLine) {
+        return requestLine[1];
+    }
 
-        fullRequest.stream()
-                .filter(line -> line.contains(":"))
-                .map(line -> split(line, ":"))
-                .forEach(line -> headersAndBody.put(line[0].trim(), line[1].trim()));
+    private String parseBody(List<String> request) {
+        String lastLine = request.get(request.size() - 1);
 
-        return headersAndBody;
+        return isBody(lastLine) ? getBodyValue(lastLine) : "";
     }
 
-    private String[] split(String request, String splitter) {
-        return request.split(splitter);
+    private HashMap<String, String> parseHeaders(List<String> request) {
+        return (HashMap<String, String>) request.stream()
+                .filter(this::isHeader)
+                .map(this::splitNameValue)
+                .collect(Collectors.toMap(line -> line.get(0), line -> line.get(1)));
     }
 
-    private String[] splitRequestLine(List<String> fullRequest) {
-        return split(fullRequest.get(0), " ");
+    private boolean isBody(String line) {
+        return line.startsWith("body");
```

Comment:
What is this method all about? How does this determine if it is the body?

Reply:
It checks to see if the line starts with "body".  I've renamed the method to `startsWithBodyTag`.  Also renamed `isHeader` to `containsHeaderDelimiter`.  Hopefully those are a little more obvious from the outside.

Reply:
🤔 The isn't guaranteed to contain a "body" though, right?

Reply:
Oh, I add it here if the request has content length [here](https://github.com/saralein/java-server/blob/response-builder/server/src/main/java/com/saralein/server/connection/ConnectionSocket.java#L37).  I was assuming that if it has content length, it has a body and it gets the "body" tag.

Reply:
Ok. I suppose I missed that. I am not sure how I feel about modifying the request body like that. Is that used anywhere except in the parser?

At any rate, there is some knowledge duplication going on here between the parser and the connection socket.

Reply:
Just in the parser.  I can rework it.